### DISCUSSION TITLE
next/dynamic with `suspense: true` only works with default exported component

### DISCUSSION THREAD

**Discussion by bryanprimus** [vercel/next.js#41149] [Help] [answered]
I got this error when using named export

![image](https://user-images.githubusercontent.com/46083126/193735938-821a4bc8-3f0e-4b57-b87d-67994bf28763.png)

![image](https://user-images.githubusercontent.com/46083126/193735845-764f9b4d-d768-4533-8627-c538e58e8628.png)

When I exported the component as default export and imported the default exported one

![image](https://user-images.githubusercontent.com/46083126/193736012-b9212a1b-843a-41f0-9fe1-949194f391c7.png)

it doesn't show the error. 

![image](https://user-images.githubusercontent.com/46083126/193736324-deb743ca-6710-4dff-814d-3b7aa280dc87.png)

I'm not sure if this is by design or not since the [docs](https://nextjs.org/docs/advanced-features/dynamic-import#example)  only show the one with default export that's why I'm not creating an issue. 

Any kind of help or answer would be really appreciated. Thanks

**Comment by icyJoseph** [comment, marked as answer]
I think when you set `suspense: true`, then the mechanism uses `React.lazy`, and that one expects the dynamic import, `() => import('something')`, within [to have a default export](https://reactjs.org/docs/code-splitting.html#reactlazy).

> React.lazy takes a function that must call a dynamic import(). This must return a Promise which resolves to a module with a default export containing a React component.

And from the next.js source code I found:

```js
 /**
       * TODO: Currently, next/dynamic will opt-in to React.lazy if { suspense: true } is used
       * React 18 will always resolve the Suspense boundary on the server-side, effectively ignoring the ssr option
       *
       * In the future, when React Suspense with third-party libraries is stable, we can implement a custom version of
       * React.lazy that can suspense on the server-side while only loading the component on the client-side
       */
```

I guess, as far as the code and the elements put to use are concerned, this is working as expected, for now.

**Reply by bryanprimus** [reply]
Thank you for the answer. @icyJoseph 

I ended up looking up how to use named export in React.Lazy and found this solution.
https://stackoverflow.com/a/66289130/12720875

Now I can make it work with that workaround in next dynamic like this

```ts
const Header = dynamic(
  () =>
    import('../components/header').then((comps) => ({ default: comps.Header })),
  { suspense: true }
)
```

assigning manually to default variable solves the error.
```ts
({ default: comps.Header })) 
```

but Still not sure this is the proper way tho.

still not clear from nextjs side

**Reply by icyJoseph** [reply]
I mean, if you come from React.lazy to `Next.js` then the named exports are strange, it really depends on how you look at it.

There's a bit of rough edges and doubts regarding Suspense, and its adoption by frameworks.

From the code, it is clear hat when using suspense Next.js uses on `React.lazy`, rather extending it as it normally does, and because of that default exports are necessary.

That being said though, I am not sure if there's any hidden trade-offs with that solution because a `Module` is an actual thing, and you are stubbing it with an object that has a readable property called `default`.

<img width="608" alt="Screenshot 2022-10-04 at 09 43 31" src="https://user-images.githubusercontent.com/21013447/193764855-dbe82b2c-7654-4996-8dbc-af7e011950dd.png">

Also, another thing to consider, is that with lazy loading you bring the entire module, even if you only load one component as you do. Independently of `default` and `named-exports` strategies.

```tsx
import { ComponentPropsWithoutRef } from "react";

export const foo = "foo";
export const bar = "bar";

export const Button = (props: ComponentPropsWithoutRef<"button">) => (
  <button {...props} />
);

export default Button;
```

> Dynamically importing Button from the above, necessarily also downloads `foo` and `bar`

Which is why it is a good idea to isolate components as much as possible, and then you can have a default export at the bottom of the file, while still having named export for the component.

```js
export const Button = () => <button></button>;
export default Button;
```

A compromise to accommodate all use cases, and it is a matter of taste and coding-style guide conformance per project.

Last but not least, magic comments can also help you when debugging in your `network` console or bundle-analyser tools:

```tsx
const DButton = dynamic(
  () => import(/* webpackChunkName: 'button' */ "../components/Button"),
  { suspense: true }
);
```

I don't think there's anything else to do but to accept that, at this point in time, with `suspense: true`, Next.js uses `React.lazy` directly, and `React.lazy` expects a default export.

**Reply by bryanprimus** [reply]
> Also, another thing to consider, is that with lazy loading you bring the entire module, even if you only load one component as you do. Independently of `default` and `named-exports` strategies.
> 
> ```tsx
> import { ComponentPropsWithoutRef } from "react";
> 
> export const foo = "foo";
> export const bar = "bar";
> 
> export const Button = (props: ComponentPropsWithoutRef<"button">) => (
>   <button {...props} />
> );
> 
> export default Button;
> ```
> 
> > Dynamically importing Button from the above, necessarily also downloads `foo` and `bar`

Isn't this already handled by webpack in tree shaking for dead code elimination?

**Reply by icyJoseph** [reply]
That's when you statically build the application, then it can of course tell that `foo` and `bar` are not needed. However, when it comes to dynamic stuff, then, as it stands, it brings on the entire module. 

There's one trick you can do, and that is, magic comments again:

```js

const DButton = dynamic(
  () =>
    import(
      /* webpackExports: ["default", "Button"], webpackChunkName: "button" */ "../components/Button"
    ),
  { suspense: true }
);
```

> The above cuts `foo` and `bar` from the `button` chunk.

Which is fairly easy to forget, and maintain, I think anyway. If I had secrets on that module, at some point in time I might end up leaking them just because a refactor renamed things and I forgot to update the magic comment.

https://webpack.js.org/blog/2020-10-10-webpack-5-release/#general-tree-shaking-improvements, but yeah it doesn't do it automatically for dynamic imports.

**Reply by bryanprimus** [reply]
> That's when you statically build the application, then it can of course tell that `foo` and `bar` are not needed. However, when it comes to dynamic stuff, then, as it stands, it brings on the entire module.

that makes sense. that is probably why my dynamic import takes longer time to be rendered. 



> There's one trick you can do, and that is, magic comments again:
> 
> ```js
> const DButton = dynamic(
>   () =>
>     import(
>       /* webpackExports: ["default", "Button"], webpackChunkName: "button" */ "../components/Button"
>     ),
>   { suspense: true }
> );
> ```
> 
> > The above cuts `foo` and `bar` from the `button` chunk.

I've seen this trick in the past when I used [cra](https://create-react-app.dev/) and [loadable-components](https://loadable-components.com/) for react router pages/routes code splitting.

Great stuff that I probably missing on. Thanks for the explanation
